File: Open-Source_Efforts/wgfast_efforts.R
# convert WGFAST json open-source file to an HTML table
# code from Colin Millar and Hjalte Parner from ICES IT
#
# to run this from the command line:
# set the working directory to the directory of the program 
# e.g., > setwd("NOAA_Gdrive/sonarpros/R_programs/WGFAST")
# source("wgfast_efforts.R")


# read in table info:
wgfast_efforts <- jsonlite::read_json("Open-Source_Efforts/WGFAST_open-source_efforts.json")

# create the rows of the table to create
# note: column names provide the table headers
table_df <-
  do.call(
    rbind,
    lapply(wgfast_efforts$efforts, function(effort) {
      data.frame(
        Title = effort$title,
        url = effort$url,
        Language = effort$programming_language,
        Description = effort$description,
        `Contact name` = effort$contact$name,
        `Contact email` = effort$contact$`e-mail`,
        check.names = FALSE
      )
    })
  )

# render the document
rmarkdown::render(
  "Open-Source_Efforts/wgfast_efforts.Rmd",
  params = list(
    title = wgfast_efforts$title,
    description = wgfast_efforts$description,
    table_df = table_df
  )
)

File: Open-Source_Efforts/wgfast_efforts.py
"""Create a .html file from the .json file."""
import json
import html
from datetime import date
from pathlib import Path
import pandas as pd
import itables

data_dir = Path('Open-Source_Efforts')
efforts_file = data_dir/'WGFAST_open-source_efforts.json'
html_file = data_dir/'wgfast_efforts.html'

with open(efforts_file, 'r', encoding='utf-8') as file:
    efforts = json.load(file)

df = pd.json_normalize(data=efforts['efforts']).drop(['contact.institution'], axis=1)
df.rename(columns={'title': 'Title', 'programming_language': 'Language',
                   'description': 'Description', 'url': 'URL', 'contact.name': 'Contact name',
                   'contact.e-mail': 'Contact e-mail'}, inplace=True)
df = df[df.Title != '']

# Make URLs proper, safe, and clickable
df['URL'] = df.apply(lambda c: '<a href="{0}">{0}</a>'
                     .format(html.escape(c['URL'], quote=True)), axis=1)
df['Contact email'] = df.apply(lambda e: '<a href="mailto:{0}">{0}</a>'
                               .format(html.escape(e['Contact e-mail'], quote=True)), axis=1)

table = itables.to_html_datatable(df, allow_html=True, paging=True, showIndex=False,
                                  footer=True, classes="display",
                                  style="width:100%")

with open(data_dir/'wgfast_efforts.html', 'w', encoding='utf-8') as f:
    f.write('<h1>' + html.escape(efforts['title']) + '</h1>')
    f.write('<p><small>Updated on ' + date.today().strftime('%d %b %Y') + '</small></p>')
    f.write('<p>' + html.escape(efforts['description']) + '</p>')
    f.write(table)
